Fix multi gpu rendering

AlexanderVeselov opened the pull request.

AlexanderVeselov committed:
Add skipping non-actual data
that would be sent from a worker device

AlexanderVeselov committed:
Add separate dirty flags to each scene controller
o Replace bool m_dirty to std::uint32_t in Baikal::SceneObject
o Each scene controller checks&resets corresponding bit in dirty mask
+ Add Scene1::Acquire and Scene1::Release methods that use mutex to
prevent data race when compiling a scene
o Fix multipass AOV crash when using mgpu config

AlexanderVeselov committed:
Some mgpu fixes
o Fix gui to show multiple devices
o Fix sample counter to account samples from worker threads
o Fix ConfigManager::CreateConfigs when it used to force disable interop even if we have interop device in mgpu
o Fix crash on exit with enabled denoiser and mgpu
+ Add thread join to AppClRender::StopRenderThreads

AlexanderVeselov committed:
Fix Ubuntu build
o Fix signed-unsigned comparison error

den13k commented on Baikal/SceneGraph/scene_object.cpp:
```diff
@@ -1,16 +1,64 @@
+
+    void SceneObject::SetSceneControllerId(std::uint32_t controller_id)
+    {
+        assert(controller_id >= 0 && controller_id < sizeof(std::uint32_t) * 8);
```
the first one is always true.
pls. remove

den13k commented on Baikal/SceneGraph/scene_object.cpp:
```diff
@@ -1,16 +1,64 @@
+    bool SceneObject::IsDirty() const
+    {
+        bool dirty = false;
+        if (g_scene_controller_id == -1)
```
is this possible?
assert should be enough

den13k commented on Baikal/SceneGraph/scene_object.cpp:
```diff
@@ -1,16 +1,64 @@
+        }
+        else
+        {
+            if (g_scene_controller_id == -1)
```
is this possible?

den13k commented on Baikal/Controllers/scene_controller.inl:
```diff
@@ -16,9 +16,13 @@
     inline
-    SceneController<CompiledScene>::SceneController() {}
+    SceneController<CompiledScene>::SceneController()
+        : m_id(g_next_scene_controller_id++)
```
does scene controllers persist while app is running? Is it possible that a controller will be deleted and the created with another id?

den13k replied:
discussed

den13k commented on Baikal/SceneGraph/scene_object.cpp:
```diff
@@ -1,16 +1,64 @@
+        }
+        else
+        {
+            dirty = ((int)m_dirty & (1 << g_scene_controller_id)) == (1 << g_scene_controller_id);
```
pls. introduce an additional variable and simplify the statement.

den13k commented on Baikal/SceneGraph/scene_object.cpp:
```diff
@@ -1,16 +1,64 @@
+            }
+            else
+            {
+                m_dirty &= ~(1 << g_scene_controller_id);
```
pls. add a comment

den13k commented:
please consider to use std::bitset

AlexanderVeselov committed:
Fix issues from the review
o Use std::bitset instead of std::uint32_t for dirty mask
o Fix asserts

den13k commented on BaikalStandalone/Application/cl_render.cpp:
```diff
@@ -54,23 +54,14 @@ namespace Baikal
-            ConfigManager::CreateConfigs(settings.mode, false, m_cfgs, settings.num_bounces, settings.platform_index, settings.device_index);
-        }
+        // Create cl context
+        ConfigManager::CreateConfigs(
```
why there is no exception handling anymore?

den13k replied:
discussed

den13k commented on Baikal/SceneGraph/scene_object.h:
```diff
@@ -58,13 +59,16 @@ namespace Baikal
         
     private:
-        mutable bool m_dirty;
+        static const int kMaxDirtyBits = 32;
```
pls add a comment here about the value

AlexanderVeselov replied:
Done

den13k commented on BaikalStandalone/Utils/config_manager.cpp:
```diff
@@ -94,6 +94,7 @@ void ConfigManager::CreateConfigs(
 
             Config cfg;
             cfg.caninterop = false;
+            bool create_secondary = true;
```
pls. rename to createWithoutInterop or something similar

AlexanderVeselov replied:
Done

AlexanderVeselov committed:
Format code
+ Add comments
+ Add empty lines
o Change variable names

AlexanderVeselov committed:
Fix Mac OSX build
o Replace CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR to explicit error code
(-1000)

AlexanderVeselov committed:
Add resetting of scene controller id in tests
